### DISCUSSION TITLE
session.bind.url.__dict__  behaviour difference in 1.3 and 1.4 version

### DISCUSSION THREAD

**Discussion by soujanyat** [sqlalchemy/sqlalchemy#11979] [Usage Questions]
I observed,

session.bind.url.__dict__  (underscore __dict__ underscore)  behaviour in 1.3 and 1.4 versions. Could you please provide more details on this and how we access the URL attributes as a dictionary in the 1.4 or 2.0 version of sqlalchemy

SQLAlchemy 1.3 version
`
sessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
>>> session = sessionLocal()
>>> session.bind.url.__dict__
{'drivername': 'mssql+pyodbc', 'username': 'user1', 'password': 'pwd', 'host': '127.X.X.X', 'port': XXX, 'database': 'database', 'query': {'driver': 'DRIVER'}}
`

SQLAlchemy 1.4 version, it is an empty dictionary
`
>>> session.bind.url.__dict__
{}
`

SQLAlchemy 2.0 version
`
>>> session.bind.url.__dict__
AttributeError: 'URL' object has no attribute '__dict__'
`

Thank you in advance for helping me understand this

**Comment by zzzeek** [comment]
the URL object in 1.4 was changed to use Python [NamedTuple](https://docs.python.org/3/library/typing.html#typing.NamedTuple) which uses [__slots__](https://docs.python.org/3/reference/datamodel.html#slots) and does not have `__dict__`.

public API for URL including access to all attributes is at https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.engine.URL

**Reply by soujanyat** [reply]
okay

Thank you

**Reply by CaselIT** [reply]
also since it's a named tuple this standard attribute should work https://docs.python.org/3/library/collections.html#collections.somenamedtuple._asdict
